File: web/lib/sidebarFeatures.ts
/**
 * Optional sidebar links (Workspace is fixed: Overview, Zilo Chat, Automations, Integrations, Features, Settings).
 * Stored at user.settings.features. Default is OFF — users turn items on under Features so the sidebar stays short.
 */

export const SIDEBAR_FEATURE_DEFAULTS: Record<string, boolean> = {
  nav_messages: false,
  nav_customers: false,
  nav_contacts: false,
  nav_suppliers: false,
  nav_investors: false,
  nav_partners: false,
  nav_followups: false,
  nav_sales: false,
  nav_orders: false,
  nav_bookings: false,
  nav_payments: false,
  nav_broadcast: false,
  nav_social_scheduler: false,
  nav_meta_ads: false,
  nav_google_ads: false,
  nav_x_ads: false,
  nav_google_business: false,
  nav_analytics: false,
  nav_team_analytics: false,
  nav_whatsapp: false,
  nav_team: false,
  nav_collaboration: false,
  nav_shop: false,
  nav_imports: false,
  nav_kds: false,
  nav_invoices: false,
  nav_inventory: false,
  nav_finance: false,
  nav_quotes: false,
  nav_loyalty: false,
  nav_nps: false,
  nav_social_inbox: false,
  nav_email: false,
  nav_email_marketing: false,
  nav_sms_marketing: false,
  nav_ai_scout: false,
  nav_behavior_tracker: false,
  nav_field_agents: false,
  nav_smart_notes: false,
  nav_calendar: false,
  nav_shopify: false,
  nav_design_templates: false,
  nav_documents: false,
  nav_seo_hub: false,
  nav_seo: false,
  nav_growth: false,
  nav_client_portal: false,
  nav_store: false,
};

/** Map route href → settings key */
export const HREF_TO_FEATURE_KEY: Record<string, string> = {
  "/dashboard/messages": "nav_messages",
  "/dashboard/customers": "nav_customers",
  "/dashboard/contacts": "nav_contacts",
  "/dashboard/suppliers": "nav_suppliers",
  "/dashboard/investors": "nav_investors",
  "/dashboard/partners": "nav_partners",
  "/dashboard/followups": "nav_followups",
  "/dashboard/sales": "nav_sales",
  "/dashboard/orders": "nav_orders",
  "/dashboard/bookings": "nav_bookings",
  "/dashboard/reservations": "nav_bookings",
  "/dashboard/payments": "nav_payments",
  "/dashboard/broadcast": "nav_broadcast",
  "/dashboard/social-scheduler": "nav_social_scheduler",
  "/dashboard/meta-ads": "nav_meta_ads",
  "/dashboard/google-ads": "nav_google_ads",
  "/dashboard/x-ads": "nav_x_ads",
  "/dashboard/google-business": "nav_google_business",
  "/dashboard/integrations": "nav_integrations",
  "/dashboard/analytics": "nav_analytics",
  "/dashboard/team-analytics": "nav_team_analytics",
  "/dashboard/whatsapp": "nav_whatsapp",
  "/dashboard/team": "nav_team",
  "/dashboard/collaboration": "nav_collaboration",
  "/dashboard/shop": "nav_shop",
  "/dashboard/imports": "nav_imports",
  "/dashboard/kds": "nav_kds",
  "/dashboard/invoices": "nav_invoices",
  "/dashboard/inventory": "nav_inventory",
  "/dashboard/finance": "nav_finance",
  "/dashboard/quotes": "nav_quotes",
  "/dashboard/loyalty": "nav_loyalty",
  "/dashboard/nps": "nav_nps",
  "/dashboard/social-inbox": "nav_social_inbox",
  "/dashboard/email": "nav_email",
  "/dashboard/email-marketing": "nav_email_marketing",
  "/dashboard/sms-marketing": "nav_sms_marketing",
  "/dashboard/action-mode": "nav_ai_scout",
  "/dashboard/marketing/behavior-discounts": "nav_behavior_tracker",
  "/dashboard/field-agents": "nav_field_agents",
  "/dashboard/smart-notes": "nav_smart_notes",
  "/dashboard/calendar": "nav_calendar",
  "/dashboard/shopify": "nav_shopify",
  "/dashboard/design-templates": "nav_design_templates",
  "/dashboard/documents": "nav_documents",
  "/dashboard/seo-hub": "nav_seo_hub",
  "/dashboard/seo": "nav_seo",
  "/dashboard/growth": "nav_growth",
  "/dashboard/client-portal": "nav_client_portal",
  "/dashboard/store": "nav_store",
};

export function mergeSidebarFeatures(raw: unknown): Record<string, boolean> {
  const out = { ...SIDEBAR_FEATURE_DEFAULTS };
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return out;
  const r = raw as Record<string, unknown>;
  for (const [k, v] of Object.entries(r)) {
    if (k in out && typeof v === "boolean") out[k] = v;
  }
  if (!out.nav_team && out.nav_collaboration) {
    out.nav_collaboration = false;
  }
  // Legacy: Team covered collaboration before nav_collaboration existed — keep both on.
  if (out.nav_team && !("nav_collaboration" in r)) {
    out.nav_collaboration = true;
  }
  // SEO Hub + SEO & Blog merged into one sidebar item — either flag turns on Website & SEO.
  if (out.nav_seo_hub === true) {
    out.nav_seo = true;
  }
  return out;
}

export function isSidebarHrefEnabled(href: string, features: Record<string, boolean>): boolean {
  if (href === "/dashboard/team" || href === "/dashboard/collaboration") {
    return true;
  }
  if (href === "/dashboard/seo") {
    return features.nav_seo === true || features.nav_seo_hub === true;
  }
  const key = HREF_TO_FEATURE_KEY[href];
  if (!key) return true;
  return features[key] === true;
}

export type FeatureToggleRow = {
  key: string;
  label: string;
  description: string;
  /** Extra terms for search (e.g. sms, text) */
  keywords?: string[];
};

export const FEATURE_TOGGLE_GROUPS: { title: string; items: FeatureToggleRow[] }[] = [
  {
    title: "Main",
    items: [
      { key: "nav_messages", label: "Messages", description: "Inbox and conversations" },
      { key: "nav_customers", label: "Customers / pipeline", description: "People & pipeline" },
      { key: "nav_contacts", label: "Contacts", description: "Contact records" },
      { key: "nav_suppliers", label: "Suppliers", description: "Vendor relationships" },
      { key: "nav_investors", label: "Investors", description: "Funding pipeline and investor relationships" },
      { key: "nav_partners", label: "Partners", description: "Strategic and channel partnerships" },
      { key: "nav_followups", label: "Follow-ups", description: "Reminders" },
    ],
  },
  {
    title: "Sales & revenue",
    items: [
      { key: "nav_sales", label: "Sales", description: "Revenue / POS tab" },
      { key: "nav_orders", label: "Orders", description: "Order list" },
      { key: "nav_bookings", label: "Bookings / Reservations", description: "Shown when your industry uses them" },
      { key: "nav_payments", label: "Payments", description: "Payment tracking" },
      { key: "nav_invoices", label: "Invoices", description: "Create and send invoices" },
      { key: "nav_quotes", label: "Quotes / Proposals", description: "Send quotes to customers" },
      { key: "nav_finance", label: "Finance / P&L", description: "Income, expenses, profit & loss" },
    ],
  },
  {
    title: "Sales & growth",
    items: [
      { key: "nav_broadcast", label: "Broadcast", description: "Campaigns to your list — retention and promos", keywords: ["whatsapp", "bulk", "promo", "campaign"] },
      { key: "nav_sms_marketing", label: "SMS Marketing", description: "Zilo notifications + your own business SMS — set up under Setup", keywords: ["sms", "text", "message", "marketing", "sender", "notification"] },
      { key: "nav_ai_scout", label: "AI Scout", description: "Autonomous agents — find leads, funding, social opportunities, and run tasks for you", keywords: ["scout", "ai", "agents", "action mode", "leads", "automation"] },
      { key: "nav_behavior_tracker", label: "Behavior Tracker", description: "GA4-powered discount campaigns triggered by visitor behavior on your site", keywords: ["behavior", "discount", "ga4", "tracking", "conversion", "cart"] },
      { key: "nav_social_scheduler", label: "Social scheduler", description: "Plan posts across networks" },
      { key: "nav_meta_ads", label: "Meta Ads", description: "Facebook & Instagram campaign planning" },
      { key: "nav_google_ads", label: "Google Ads", description: "Search & Performance Max planning" },
      { key: "nav_x_ads", label: "X Ads", description: "Promoted posts & campaigns on X (Twitter)" },
      { key: "nav_google_business", label: "Google Business Profile", description: "Maps & local presence via Integrations" },
      { key: "nav_social_inbox", label: "Social Inbox", description: "Unified DMs from all your connected social platforms" },
      { key: "nav_seo", label: "SEOhub", description: "Coach, keywords, content, autoblog site, and audits — one workspace" },
    ],
  },
  {
    title: "Business",
    items: [
      { key: "nav_analytics", label: "Analytics", description: "Dashboard metrics" },
      { key: "nav_team_analytics", label: "Team analytics", description: "Team performance" },
      { key: "nav_whatsapp", label: "WhatsApp", description: "WA tools" },
      { key: "nav_field_agents", label: "Field Agents", description: "Assign and track field rep tasks — check-ins, routes, and activity", keywords: ["field", "agents", "reps", "tasks", "check-in", "routes", "mobile"] },
      { key: "nav_smart_notes", label: "Zilo Notetaker", description: "AI meeting note-taker — auto-records and transcribes calendar meetings", keywords: ["notes", "meeting", "transcribe", "record", "notetaker", "ai"] },
      { key: "nav_team", label: "Team", description: "Members, roles, and team settings" },
      {
        key: "nav_collaboration",
        label: "Collaboration",
        description: "Workspaces, channel access, and routing — turns on automatically when Team is on",
      },
      { key: "nav_shop", label: "Shop / catalog", description: "Storefront (label varies by type)" },
      { key: "nav_imports", label: "Imports", description: "Bulk upload" },
      { key: "nav_inventory", label: "Inventory / Stock", description: "Track products and stock levels" },
      { key: "nav_loyalty", label: "Customer Loyalty", description: "Points and rewards program" },
      { key: "nav_nps", label: "Customer Feedback / NPS", description: "Surveys and satisfaction scores" },
    ],
  },
  {
    title: "Productivity",
    items: [
      { key: "nav_email", label: "Email Inbox", description: "Gmail or Outlook inbox with AI draft & auto-reply", keywords: ["gmail", "outlook", "inbox"] },
      { key: "nav_email_marketing", label: "Email Marketing", description: "Create and send email campaigns to your contacts", keywords: ["email", "newsletter", "campaign", "mail"] },
      { key: "nav_calendar", label: "Calendar", description: "Google or Outlook calendar with event management" },
      { key: "nav_shopify", label: "Shopify", description: "Orders, inventory, customers, abandoned carts & discounts" },
    { key: "nav_design_templates", label: "Design library", description: "Chat-generated graphics, PDFs, and decks plus optional manual template metadata" },
    { key: "nav_documents", label: "Documents", description: "AI-generated PDFs, Word docs, and presentations from Zilo Chat" },
    { key: "nav_growth", label: "Growth Suite", description: "Daily digest, autopilot follow-ups, deal room, competitor intel, voice input" },
    { key: "nav_client_portal", label: "Client Portal", description: "Share a portal link with clients to view their invoices, proposals & orders" },
    { key: "nav_store", label: "My Store", description: "Manage your WooCommerce store — products, orders, and payments" },
    ],
  },
  {
    title: "Display",
    items: [{ key: "nav_kds", label: "KDS display", description: "Kitchen screen (F&B types)" }],
  },
];

/** All toggle rows flat — useful for search / counts */
export function getAllFeatureToggleRows(): (FeatureToggleRow & { group: string })[] {
  return FEATURE_TOGGLE_GROUPS.flatMap((group) =>
    group.items.map((row) => ({ ...row, group: group.title }))
  );
}

/** Filter toggle groups by search query (label, description, group, keywords). */
export function filterFeatureToggleGroups(
  query: string,
  groups: { title: string; items: FeatureToggleRow[] }[] = FEATURE_TOGGLE_GROUPS
): { title: string; items: FeatureToggleRow[] }[] {
  const q = query.trim().toLowerCase();
  if (!q) return groups;
  const terms = q.split(/\s+/).filter(Boolean);
  return groups
    .map((group) => ({
      ...group,
      items: group.items.filter((row) => {
        const haystack = [
          row.label,
          row.description,
          group.title,
          ...(row.keywords ?? []),
        ]
          .join(" ")
          .toLowerCase();
        return terms.every((term) => haystack.includes(term));
      }),
    }))
    .filter((g) => g.items.length > 0);
}

/** Business preset — sales, customers, invoices, analytics */
export const PRESET_BUSINESS: Partial<Record<string, boolean>> = {
  nav_messages: true,
  nav_customers: true,
  nav_contacts: true,
  nav_suppliers: false,
  nav_followups: true,
  nav_sales: true,
  nav_orders: true,
  nav_bookings: false,
  nav_payments: true,
  nav_invoices: true,
  nav_quotes: true,
  nav_finance: true,
  nav_broadcast: true,
  nav_social_scheduler: false,
  nav_meta_ads: false,
  nav_google_ads: false,
  nav_x_ads: false,
  nav_google_business: false,
  nav_social_inbox: false,
  nav_analytics: true,
  nav_team_analytics: false,
  nav_whatsapp: true,
  nav_team: true,
  nav_collaboration: true,
  nav_shop: false,
  nav_imports: false,
  nav_inventory: false,
  nav_loyalty: false,
  nav_nps: false,
  nav_kds: false,
};

/** Personal preset — social, scheduling, broadcast */
export const PRESET_PERSONAL: Partial<Record<string, boolean>> = {
  nav_messages: true,
  nav_customers: false,
  nav_contacts: true,
  nav_suppliers: false,
  nav_followups: true,
  nav_sales: false,
  nav_orders: false,
  nav_bookings: false,
  nav_payments: false,
  nav_invoices: false,
  nav_quotes: false,
  nav_finance: false,
  nav_broadcast: true,
  nav_social_scheduler: true,
  nav_meta_ads: false,
  nav_google_ads: false,
  nav_social_inbox: true,
  nav_analytics: false,
  nav_team_analytics: false,
  nav_whatsapp: true,
  nav_team: true,
  nav_collaboration: true,
  nav_shop: false,
  nav_imports: false,
  nav_inventory: false,
  nav_loyalty: false,
  nav_nps: false,
  nav_kds: false,
};

/** Quick-add bundle: common tools without turning everything on */
export const PRESET_STARTER: Partial<Record<string, boolean>> = {
  nav_messages: true,
  nav_customers: true,
  nav_contacts: false,
  nav_suppliers: false,
  nav_followups: true,
  nav_sales: false,
  nav_orders: false,
  nav_bookings: false,
  nav_payments: false,
  nav_broadcast: false,
  nav_social_scheduler: false,
  nav_meta_ads: false,
  nav_google_ads: false,
  nav_x_ads: false,
  nav_google_business: false,
  nav_analytics: false,
  nav_team_analytics: false,
  nav_whatsapp: true,
  nav_team: true,
  nav_collaboration: true,
  nav_shop: false,
  nav_imports: false,
  nav_kds: false,
};

/** Every optional key set from a partial (defaults for missing keys = `defaultValue`, usually false). */
export function applyAllFromPartial(preset: Partial<Record<string, boolean>>, defaultValue = false): Record<string, boolean> {
  const out = { ...SIDEBAR_FEATURE_DEFAULTS };
  for (const k of Object.keys(out)) {
    out[k] = k in preset ? Boolean(preset[k]) : defaultValue;
  }
  if (out.nav_team && !("nav_collaboration" in preset)) {
    out.nav_collaboration = true;
  }
  if (!out.nav_team) {
    out.nav_collaboration = false;
  }
  return out;
}

/** Enable every optional sidebar link (full workspace). */
export function allSidebarFeaturesOn(): Record<string, boolean> {
  const out = { ...SIDEBAR_FEATURE_DEFAULTS };
  for (const k of Object.keys(out)) out[k] = true;
  return out;
}

File: backend/assistant/sidebar_features.py
"""
sidebar_features.py
Sidebar feature catalog + contextual recommendations for the assistant.

Used when the user's goal requires a CRM module that is not enabled yet.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

# feature_key → metadata (keep in sync with web/lib/sidebarFeatures.ts)
FEATURE_CATALOG: Dict[str, Dict[str, str]] = {
    "nav_messages": {"label": "Messages", "path": "/dashboard/messages", "group": "Main"},
    "nav_customers": {"label": "Customers / pipeline", "path": "/dashboard/customers", "group": "Main"},
    "nav_contacts": {"label": "Contacts", "path": "/dashboard/contacts", "group": "Main"},
    "nav_followups": {"label": "Follow-ups", "path": "/dashboard/followups", "group": "Main"},
    "nav_sales": {"label": "Sales", "path": "/dashboard/sales", "group": "Sales & revenue"},
    "nav_orders": {"label": "Orders", "path": "/dashboard/orders", "group": "Sales & revenue"},
    "nav_bookings": {"label": "Bookings", "path": "/dashboard/bookings", "group": "Sales & revenue"},
    "nav_invoices": {"label": "Invoices", "path": "/dashboard/invoices", "group": "Sales & revenue"},
    "nav_quotes": {"label": "Quotes", "path": "/dashboard/quotes", "group": "Sales & revenue"},
    "nav_finance": {"label": "Finance / P&L", "path": "/dashboard/finance", "group": "Sales & revenue"},
    "nav_broadcast": {"label": "Broadcast", "path": "/dashboard/broadcast", "group": "Sales & growth"},
    "nav_sms_marketing": {"label": "SMS Marketing", "path": "/dashboard/sms-marketing", "group": "Sales & growth"},
    "nav_ai_scout": {"label": "AI Scout", "path": "/dashboard/action-mode", "group": "Sales & growth"},
    "nav_behavior_tracker": {"label": "Behavior Tracker", "path": "/dashboard/marketing/behavior-discounts", "group": "Sales & growth"},
    "nav_social_scheduler": {"label": "Social scheduler", "path": "/dashboard/social-scheduler", "group": "Sales & growth"},
    "nav_meta_ads": {"label": "Meta Ads", "path": "/dashboard/meta-ads", "group": "Sales & growth"},
    "nav_google_ads": {"label": "Google Ads", "path": "/dashboard/google-ads", "group": "Sales & growth"},
    "nav_x_ads": {"label": "X Ads", "path": "/dashboard/x-ads", "group": "Sales & growth"},
    "nav_social_inbox": {"label": "Social Inbox", "path": "/dashboard/social-inbox", "group": "Sales & growth"},
    "nav_seo": {"label": "SEOhub", "path": "/dashboard/seo", "group": "Sales & growth"},
    "nav_whatsapp": {"label": "WhatsApp", "path": "/dashboard/whatsapp", "group": "Business"},
    "nav_field_agents": {"label": "Field Agents", "path": "/dashboard/field-agents", "group": "Business"},
    "nav_team": {"label": "Team", "path": "/dashboard/team", "group": "Business"},
    "nav_collaboration": {"label": "Collaboration", "path": "/dashboard/collaboration", "group": "Business"},
    "nav_analytics": {"label": "Analytics", "path": "/dashboard/analytics", "group": "Business"},
    "nav_inventory": {"label": "Inventory", "path": "/dashboard/inventory", "group": "Business"},
    "nav_loyalty": {"label": "Loyalty", "path": "/dashboard/loyalty", "group": "Business"},
    "nav_nps": {"label": "Feedback / NPS", "path": "/dashboard/nps", "group": "Business"},
    "nav_email": {"label": "Email Inbox", "path": "/dashboard/email", "group": "Productivity"},
    "nav_email_marketing": {"label": "Email Marketing", "path": "/dashboard/email-marketing", "group": "Productivity"},
    "nav_calendar": {"label": "Calendar", "path": "/dashboard/calendar", "group": "Productivity"},
    "nav_shopify": {"label": "Shopify", "path": "/dashboard/shopify", "group": "Productivity"},
    "nav_kds": {"label": "KDS display", "path": "/dashboard/kds", "group": "Display"},
}

# When user intent maps to a feature (keyword → feature keys to check)
INTENT_TO_FEATURES: Dict[str, List[str]] = {
    "sms": ["nav_sms_marketing"],
    "text message": ["nav_sms_marketing"],
    "broadcast": ["nav_broadcast"],
    "whatsapp promo": ["nav_broadcast", "nav_whatsapp"],
    "field rep": ["nav_field_agents"],
    "field agent": ["nav_field_agents"],
    "check-in": ["nav_field_agents"],
    "discount website": ["nav_behavior_tracker"],
    "cart abandon": ["nav_behavior_tracker"],
    "behavior track": ["nav_behavior_tracker"],
    "scout": ["nav_ai_scout"],
    "find leads": ["nav_ai_scout"],
    "social inbox": ["nav_social_inbox"],
    "dm": ["nav_social_inbox"],
    "schedule post": ["nav_social_scheduler"],
    "facebook ad": ["nav_meta_ads"],
    "instagram ad": ["nav_meta_ads"],
    "google ad": ["nav_google_ads"],
    "invoice": ["nav_invoices"],
    "quote": ["nav_quotes"],
    "loyalty": ["nav_loyalty"],
    "nps": ["nav_nps"],
    "feedback survey": ["nav_nps"],
    "email campaign": ["nav_email_marketing"],
    "seo": ["nav_seo"],
    "blog": ["nav_seo"],
    "inventory": ["nav_inventory"],
    "stock": ["nav_inventory"],
    "kitchen display": ["nav_kds"],
    "kds": ["nav_kds"],
    "booking": ["nav_bookings"],
    "reservation": ["nav_bookings"],
    "team member": ["nav_team"],
    "collaboration": ["nav_collaboration"],
}

# Priority picks by business type (for open-ended "what should I use?" — max 3-4)
RECOMMENDED_BY_TYPE: Dict[str, List[str]] = {
    "retail": ["nav_customers", "nav_whatsapp", "nav_broadcast", "nav_invoices", "nav_inventory"],
    "wholesale": ["nav_customers", "nav_orders", "nav_invoices", "nav_quotes", "nav_field_agents"],
    "restaurant": ["nav_orders", "nav_bookings", "nav_kds", "nav_whatsapp", "nav_broadcast"],
    "food": ["nav_orders", "nav_whatsapp", "nav_broadcast"],
    "bakery": ["nav_orders", "nav_bookings", "nav_whatsapp", "nav_broadcast"],
    "grocery": ["nav_orders", "nav_inventory", "nav_whatsapp", "nav_loyalty"],
    "hotel": ["nav_bookings", "nav_customers", "nav_invoices", "nav_followups"],
    "rental": ["nav_bookings", "nav_customers", "nav_invoices", "nav_quotes"],
    "salon": ["nav_bookings", "nav_customers", "nav_followups", "nav_whatsapp", "nav_broadcast"],
    "spa": ["nav_bookings", "nav_customers", "nav_followups", "nav_whatsapp"],
    "fitness": ["nav_bookings", "nav_customers", "nav_loyalty", "nav_broadcast"],
    "healthcare": ["nav_bookings", "nav_customers", "nav_followups", "nav_calendar"],
    "services": ["nav_quotes", "nav_invoices", "nav_followups", "nav_field_agents"],
    "repair": ["nav_orders", "nav_field_agents", "nav_quotes", "nav_whatsapp"],
    "cleaning": ["nav_field_agents", "nav_bookings", "nav_invoices", "nav_whatsapp"],
    "events": ["nav_quotes", "nav_invoices", "nav_calendar", "nav_social_scheduler"],
    "creator": ["nav_social_inbox", "nav_social_scheduler", "nav_broadcast", "nav_email_marketing"],
    "support": ["nav_messages", "nav_team", "nav_nps", "nav_followups"],
    "general": ["nav_customers", "nav_whatsapp", "nav_followups", "nav_analytics"],
}

FEATURES_PAGE = "/dashboard/features"

# After a tool succeeds, suggest enabling a sidebar module if still off (one nudge per tool).
POST_ACTION_NUDGES: Dict[str, Dict[str, str]] = {
    "create_scheduled_post": {
        "feature_key": "nav_social_scheduler",
        "reason": "You just saved a post — turn on **Social scheduler** to manage your content calendar from the sidebar.",
    },
    "generate_social_post": {
        "feature_key": "nav_social_scheduler",
        "reason": "Your post creative is ready — enable **Social scheduler** in Features to schedule and track posts from the sidebar.",
    },
    "generate_ad_creative": {
        "feature_key": "nav_social_scheduler",
        "reason": "Your creative is ready — **Social scheduler** lets you queue posts and plan your calendar from the sidebar.",
    },
    "create_broadcast": {
        "feature_key": "nav_broadcast",
        "reason": "To run WhatsApp campaigns from the sidebar anytime, turn on **Broadcast** under Features.",
    },
    "create_followup": {
        "feature_key": "nav_followups",
        "reason": "You set a follow-up — enable **Follow-ups** in Features to see and manage reminders in your sidebar.",
    },
    "create_customer": {
        "feature_key": "nav_customers",
        "reason": "You added a contact — turn on **Customers / pipeline** in Features for full CRM access in the sidebar.",
    },
}


def post_action_nudge(
    tool_name: str,
    tool_result: Any,
    features: Dict[str, Any],
) -> Optional[Dict[str, Any]]:
    """
    Return a single contextual nudge after a successful tool run, or None.
    Only when the related sidebar feature is not enabled yet.
    """
    if not isinstance(tool_result, dict) or tool_result.get("error"):
        return None
    if tool_result.get("success") is False:
        return None

    spec = POST_ACTION_NUDGES.get(tool_name)
    if not spec:
        return None

    key = spec["feature_key"]
    if _is_enabled(features, key):
        return None

    row = _feature_row(
        key,
        reason=spec["reason"],
        enabled=False,
    )
    if not row:
        return None

    row["trigger_tool"] = tool_name
    row["when_to_say"] = (
        "After confirming what you just did for the user, add ONE short sentence: "
        f"for ongoing access they should enable {row['label']} via Features (search → toggle). "
        "Do not pitch other modules."
    )
    return row


def enrich_tool_result_with_nudge(
    tool_name: str,
    tool_result: Any,
    features: Dict[str, Any],
) -> Any:
    """Attach `_feature_hint` to tool JSON the model sees after successful actions."""
    nudge = post_action_nudge(tool_name, tool_result, features)
    if not nudge or not isinstance(tool_result, dict):
        return tool_result
    out = dict(tool_result)
    out["_feature_hint"] = nudge
    return out


def _normalize_type(business_type: str) -> str:
    t = (business_type or "general").strip().lower()
    if t in RECOMMENDED_BY_TYPE:
        return t
    aliases = {
        "beauty": "salon",
        "ecommerce": "retail",
        "shop": "retail",
        "cafe": "restaurant",
        "hospitality": "hotel",
    }
    return aliases.get(t, "general")


def _is_enabled(features: Dict[str, Any], key: str) -> bool:
    return features.get(key) is True


def _feature_row(key: str, *, reason: str, enabled: bool) -> Optional[Dict[str, Any]]:
    meta = FEATURE_CATALOG.get(key)
    if not meta:
        return None
    return {
        "key": key,
        "label": meta["label"],
        "path": meta["path"],
        "group": meta["group"],
        "enabled": enabled,
        "reason": reason,
        "enable_instructions": (
            f"Go to **Features** ({FEATURES_PAGE}), search for \"{meta['label']}\", and turn the toggle on."
        ),
    }


def features_for_intent(user_message: str, features: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Return disabled features required for the user's stated intent."""
    msg = (user_message or "").lower()
    matched_keys: List[str] = []
    for phrase, keys in INTENT_TO_FEATURES.items():
        if phrase in msg:
            for k in keys:
                if k not in matched_keys:
                    matched_keys.append(k)

    out: List[Dict[str, Any]] = []
    for key in matched_keys:
        if _is_enabled(features, key):
            continue
        meta = FEATURE_CATALOG.get(key, {})
        reason = f"You need **{meta.get('label', key)}** to do this in Zilo."
        row = _feature_row(key, reason=reason, enabled=False)
        if row:
            out.append(row)
    return out[:3]


def recommendations_for_business(
    business_type: str,
    features: Dict[str, Any],
    *,
    limit: int = 4,
) -> List[Dict[str, Any]]:
    """Suggest features commonly useful for this business type that are not enabled yet."""
    btype = _normalize_type(business_type)
    keys = RECOMMENDED_BY_TYPE.get(btype, RECOMMENDED_BY_TYPE["general"])
    out: List[Dict[str, Any]] = []
    for key in keys:
        if _is_enabled(features, key):
            continue
        meta = FEATURE_CATALOG.get(key, {})
        label = meta.get("label", key)
        reason = f"Common for {btype.replace('_', ' ')} businesses like yours — **{label}** helps with day-to-day work."
        row = _feature_row(key, reason=reason, enabled=False)
        if row:
            out.append(row)
        if len(out) >= limit:
            break
    return out


def build_feature_guidance(
    *,
    business_type: str,
    features: Dict[str, Any],
    user_intent: str = "",
    mode: str = "intent",
    limit: int = 3,
) -> Dict[str, Any]:
    """
    mode:
      - intent: features required for user_intent (task-specific)
      - profile: business-type suggestions (only when user asks what to enable)
    """
    enabled_count = sum(1 for k in FEATURE_CATALOG if _is_enabled(features, k))
    recommendations: List[Dict[str, Any]] = []

    if mode == "profile":
        recommendations = recommendations_for_business(business_type, features, limit=limit)
    elif mode == "intent" and user_intent.strip():
        recommendations = features_for_intent(user_intent, features)

    return {
        "business_type": _normalize_type(business_type),
        "features_page": FEATURES_PAGE,
        "enabled_sidebar_tools": enabled_count,
        "total_optional_tools": len(FEATURE_CATALOG),
        "recommendations": recommendations[:limit],
        "guidance_tone": (
            "Only mention a feature when the user's current goal requires it and it is disabled. "
            "Say what they need it for, then: go to Features, search the tool name, toggle on. "
            "Do not list unrelated tools or repeat after they decline."
        ),
    }
